File: Medium/216. Combination Sum III.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> list:
        self.ans = set()
        self.dfs(k, n, [])
        return [x for x in self.ans]
        
    def dfs(self, k, curNum, usedNum:list):
        if k == 1:
            if curNum not in usedNum and 0 < curNum < 10:
                usedNum.append(curNum)
                usedNum.sort()
                self.ans.add(tuple(usedNum))
        else:
            start = max(usedNum)+1 if usedNum else 1
            for i in range(start, 10):
                if curNum - i > 0:
                    passNums = usedNum.copy()
                    passNums.append(i)
                    self.dfs(k-1, curNum-i, passNums)

File: Medium/216. Combination Sum III.cpp
class Solution {
public:
	vector<vector<int>> ans;
	
    vector<vector<int>> combinationSum3(int k, int n) {
        dfs(k, n, vector<int>(), 1);
		return ans;
    }
	
	void dfs(int k, int curNum, vector<int> usedNum, int start){
		if(k == 0){
			if(curNum == 0){
				ans.push_back(usedNum);
			}
		}else{
			for(int i = start; i<10; i++){
				if(curNum - i >= 0){
					usedNum.push_back(i);
					dfs(k-1, curNum-i, usedNum, i+1);
					usedNum.pop_back();
				}
			}
		}
	}
};